cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/CoreSchedulerService.java

```diff
@@ -188,11 +191,37 @@ private void migrateSchedules(final NamespaceQueryAdmin namespaceQueryAdmin, fin
     LOG.info("Schedule migration is completed for all namespaces.");
   }
 
+  /**
+   * Waits for this scheduler completely started and functional.
+   *
+   * @param timeout maximum timeout to wait for
+   * @param timeoutUnit unit for the timeout
+   *
+   * @throws InterruptedException if the current thread is being interrupted
+   * @throws TimeoutException if the scheduler is not yet functional until the give timeout time passed
+   * @throws IllegalStateException if this scheduler is not yet started
+   */
+  @VisibleForTesting
+  public void waitUntilFunctional(long timeout, TimeUnit timeoutUnit) throws TimeoutException, InterruptedException {
+    if (!isRunning()) {
+      throw new IllegalStateException(getClass().getSimpleName()
+                                        + " is not running. Cannot wait for it to be functional.");
+    }
+    if (!startedLatch.await(timeout, timeoutUnit)) {
+      throw new TimeoutException(getClass().getSimpleName()
+                                   + " is not completely started after " + timeout + " " + timeoutUnit);
+    }
+  }
+
+  /**
+   * Checks if the scheduler started completely and is functional.
+   *
+   * @throws ServiceUnavailableException if the scheduler is not yet functional
+   */
   private void checkStarted() {
-    if (schedulerStarted.get()) {
-      return;
+    if (!Uninterruptibles.awaitUninterruptibly(startedLatch, 0, TimeUnit.SECONDS)) {
```

Comment:
what is the purpose of calling `Uninterruptibles.awaitUninterruptibly` with 0 sec timeout?

Reply:
We don't want the interrupted exception to be thrown in this case, as we just want to check. Calling with a `0` timeout means if won't wait for the latch if the count is not 0.